packages/route/src/Infrastructure/Doctrine/EventListener/RouteChangedUpdater.php

```diff
@@ -0,0 +1,124 @@
+        ];
+    }
+
+    public function postFlush(PostFlushEventArgs $args): void
+    {
+        if (0 === \count($this->routeChanges)) {
+            return;
+        }
+
+        $connection = $args->getObjectManager()->getConnection();
+
+        $routesTableName = $args->getObjectManager()->getClassMetadata(Route::class)->getTableName();
+
+        foreach ($this->routeChanges as $routeChange) {
+            $oldSlug = $routeChange['oldValue'];
+            $newSlug = $routeChange['newValue'];
+            $locale = $routeChange['locale'];
+            $site = $routeChange['site'];
+
+            // select all child and grand routes of oldSlug
+            $selectQueryBuilder = $connection->createQueryBuilder()
+                ->from($routesTableName, 'parent')
+                ->select('parent.id AS parent_id')
+                ->innerJoin('parent', $routesTableName, 'child', 'child.parent_id = parent.id')
+                ->andWhere(\is_string($site) ? 'parent.site = :site' : 'parent.site IS NULL')
+                ->andWhere('parent.locale = :locale')
+                ->andWhere('(parent.slug = :newSlug OR parent.slug LIKE :oldSlugSlash)') // direct child is using newSlug already updated as we are in PostFlush, grand child use oldSlugWithSlash as not yet updated
+                ->setParameter('newSlug', $newSlug, ParameterType::STRING)
+                ->setParameter('oldSlugSlash', $oldSlug . '/%', ParameterType::STRING)
+                ->setParameter('locale', $locale, ParameterType::STRING);
+
+            if (\is_string($site)) {
+                $selectQueryBuilder->setParameter('site', $site, ParameterType::STRING);
+            }
+
+            $parentIds = \array_map(fn ($row) => $row[0], $selectQueryBuilder->executeQuery()->fetchAllNumeric());
+            $parentIds = \array_filter($parentIds);
+
+            if (0 === \count($parentIds)) {
+                continue;
+            }
+
+            $parentIds = \array_unique($parentIds); // DISTINCT and GROUP BY a lot slower as make it unique in PHP itself
+
+            // TODO create history for current ids
+
+            $newSlugCast = '';
+            if ($connection->getDatabasePlatform() instanceof PostgreSQLPlatform) {
+                $newSlugCast = '::text'; // concat seems not directly supported by dbal and parameter $1 (newSlug) is not cast to text correctly. So manually cast it here.
+            }
+
+            // update child and grand routes
+            $updateQueryBuilder = $connection->createQueryBuilder()
+                ->update($routesTableName, 'r')
+                ->set('slug', 'CONCAT(:newSlug' . $newSlugCast . ', SUBSTRING(slug, ' . (\strlen($oldSlug) + 1) . '))')
+                ->setParameter('newSlug', $newSlug, ParameterType::STRING)
+                ->where('parent_id IN (:parentIds)')
+                ->setParameter('parentIds', $parentIds, ArrayParameterType::INTEGER);
+
+            $updateQueryBuilder->executeStatement();
```

Comment:
Not ideal handling of Postgres but works now as expected and is covers by tests.